Include index in generated profile name

codeinred opened the pull request:
This PR updates generate.sh to include the index of the theme in the display name of the theme recorded in themes.json.

Having the index of the theme as part of the display name will make installation easier, since users will be able to see the number they're supposed to enter in order to install a theme listed as part of the theme name on <https://mayccoll.github.io/Gogh/>.

The name is now listed as: `<Theme Name> (theme ##)`

Mayccoll commented:
Hi @codeinred  the issue with this approach is that "index" is going to change every time someone add a new theme, and if someone create an app or some script based on that index later the same index is going to match with other theme.
I don't know if it's clear?
For example, someone create an script that specifically install the` index = 110 `then when Gogh updates that 110 is going to point to another theme.
I want to solve this creating a index width a fixed number, but this is not scalable if done manually.

For me the best approach is create some hash number based on the theme name and good  enough that it avoid collisions

something like this:

`echo $((0x$(echo 'THEME NAME' | md5sum | cut -f 1 -d " " | cut -c 1-9))) | cut -c 1-9`

This example always is going to generate this number `326432603` i know is long for typing that why i like the consecutive number but for scripts as an `ID` is going to be perfect.

also can be as simple as `echo 'THEME NAME' | md5sum and grap the first 5 digits, but i want to avoid letters.
`799b0`

What do you think about it and what do you think is the best approach?

Thank you
`

codeinred commented:
Hi @Mayccoll, 

Thank you so much for taking the time to review this PR and to write out your reasoning! I understand the concern about indices changing, especially since it could lead someone to try relying on it for a script, resulting in a surprise when they got shifted after the addition of a new theme.

What if we enabled the install script to look up themes by name, rather than by index, with themes separated by commas? This should be possible with xargs. Then, based on the theme name, we could look up the file corresponding to the theme. This would potentially require renaming some of the files, but we can preserve history with `git mv` and automate any renaming with a script.

So if user entered `Theme Name Light, Theme Name Dark`, it would translate that to `(theme-name-light.shtheme-name-dark.sh)`

Mayccoll commented:
:smile:  
The name can lead to misspelling.

 what about to keep the number for install manually in console, like it is right now, and with the option for ID.

So if people write 9 numbers it will use the ID and if they write less that those digits it will use the consecutive number.
Also the theme name for install can be added but that's is more coding :) .

The json will be something like this.

```
{
      "id": "411355344",
      "name": "3024Day",
      "black": "#090300",
      "red": "#db2d20",
      "green": "#01a252",
      "yellow": "#fded02",
      "blue": "#01a0e4",
      "purple": "#a16a94",
      "cyan": "#b5e4f4",
      "white": "#a5a2a2",
      "brightBlack": "#5c5855",
      "brightRed": "#e8bbd0",
      "brightGreen": "#3a3432",
      "brightYellow": "#4a4543",
      "brightBlue": "#807d7c",
      "brightPurple": "#d6d5d4",
      "brightCyan": "#cdab53",
      "brightWhite": "#f7f7f7",
      "foreground": "#4a4543",
      "background": "#f7f7f7",
      "cursorColor": "#4a4543"
    }
```

I haven't really had much time to work on this, I wanted to make those changes for the next big version of the GOGH and a couple of more :yum: .

My idea for the next version is that the themes will be made in separete Yaml files and everything will start from there. With that it will be even easier to contribute themes and port the themes to other languages (bash, node, python , etc).

hahaha sorry, it's a lot of plans.

Translated with www.DeepL.com/Translator (free version)

codeinred commented:
I like the idea of moving theme files to yaml in a future version! It will be good to make themes more portable, and I think yaml has a simpler and more approachable syntax than bash does. It should also be easier to export themes from other places, to add them to gogh. 

## Motivation
The reason I suggested allowing the use of names is because the current workflow feels cumbersome: 

1. Run Gogh
2. Search for themes I like on <https://mayccoll.github.io/Gogh/>
3. Find the corresponding number for each theme, based on the name
4. Enter the numbers

The trouble lies in switching repeatedly between steps 2, 3, and 4 to install multiple themes. I have to scroll through a large list of names to find the corresponding number, enter it, then scroll to find the next number. To make the process faster I sometimes try juggling multiple numbers in my head at once, but that can be error-prone.

Eliminating step 3 was the original purpose of this PR, so that numbers would show up on the website itself, next to the corresponding theme, and I could enter those without scrolling up to find the number. You made a good point about numbers changing as new themes are added, though, so it makes sense why you wouldn't want to do this.

Numeric IDs would partially solve this problem: we could list the ID on the website, next to the theme name. But an ID isn't human readable, and I would argue that typing in a long ID is more error prone than typing in a human-readable name. 

The theme names are already unique; they're human readable; and with a small tweak to the generate script, we could preserve spaces in the name so they can be copied from the website.

## Implementation details

I know I'm suggesting a large change, but I'm fully happy to implement it myself.

**Preserving spaces in the names listed on the website:** Right now all whitespace is removed from the json file using `tr` as part of a process to remove the trailing comma. Below, I've provided an alternative implementation that will remove this trailing comma without removing the whitespace:
```bash
# Use sed to remove the final trailing comma in the json file, 
# without removing spaces or newlines
cp $SCRIPT_PATH/../data/themes.json $SCRIPT_PATH/../data/themes.json.tmp
sed -e ':a' -e 'N' -e '$!ba' -e 's/},\n  ]/}\n  ]/g' \
  $SCRIPT_PATH/../data/themes.json.tmp > $SCRIPT_PATH/../data/themes.json
rm $SCRIPT_PATH/../data/themes.json.tmp
```

**Looking up files based on the name a user enters:** Below I've provided a snippet which normalizes user input, converting a list of comma-separated names into theme filenames. (When things are converted to yaml, this should be really simple to do in python just by calling replace and trim functions)

This does the following steps:
- replace commas with newlines
- Replace spaces with dashes
- Remove leading dashes
- Remove trailing dashes
- Add a .sh at the end
- Convert everything to lowercase
```bash
function getThemeFiles() {
    read themes
    echo $themes            \
        | tr ',' '\n'       \
        | tr ' ' '-'        \
        | sed -e 's/^-*//'  \
              -e 's/-*$//'  \
              -e 's/$/.sh/' \
        | tr '[:upper:]' '[:lower:]'
}
```
![image](https://user-images.githubusercontent.com/36810712/140230746-88d2c8ac-8623-49e5-83bf-85e04552be7b.png)

### Concerns regarding naming

Most but not all files follow the naming convention described above, but the ones that don't can be renamed with the following script. This will extract profile names without installing the theme.

```bash
for file in *.sh; do
    source <(head -n 30 $file)
    dest=$(echo $PROFILE_NAME | getThemeFiles)
    if [ "$file" != "$dest" ]; then 
        git mv $file $dest 
    fi
done
```

After experimenting with this, the following 9 files get renamed, based on the profile name:
```
	renamed:    gogh.sh -> _bash.sh
	renamed:    adventuretime.sh -> adventure-time.sh
	renamed:    cobalt2.sh -> cobalt-2.sh
	renamed:    fairyflossdark.sh -> fairy-floss-dark.sh
	renamed:    fairyfloss.sh -> fairy-floss.sh
	renamed:    ibm3270-hicontrast.sh -> ibm-3270-(high-contrast).sh
	renamed:    sweet.sh -> sweet-terminal.sh
	renamed:    vs-code-dark-plus.sh -> vs-code-dark+.sh
	renamed:    vs-code-light-plus.sh -> vs-code-light+.sh
```

Mayccoll commented:
I definitely like the idea of being able to install Gogh with names also separated by commas.

What I'm not so sure about is whats going to be the source of truth and it's something I've been struggling with since I started this application. The other day I lost about 2 hours because I had a bad typo in the filename and in the `gogh.sh` file.

I don't think we can use the file name as source of truth, which is why I have a hard time with this approach.

Maybe using a field `export FINENAME="adventure-time.sh"` in combination with the API

`https://api.github.com/repos/Mayccoll/Gogh/git/trees/master?recursive=1`

I know right now the source of truth for the command line are the file names, but for the website is the field.

My idea at the end (next version) is use **Actions** to generate everything ones you add a new `YAML` file, righ now you need to modify the `gogh.sh` file adding the new theme to the list.

i don't know, need to think a little more :)

I send you and email with my discord. or let me know if you want to contact me through other means. It's hard to write so much :smile:
